### DISCUSSION TITLE
Most elegant way to serve only a subset of data/endpoints of a larger backend

### DISCUSSION THREAD

**Discussion by sla-te** [fastapi/fastapi#10780] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Operating System

Linux

### Operating System Details

ubuntu server

### FastAPI Version

0.101.1

### Pydantic Version

1.10.2

### Python Version

3.11.4

### Additional Context

Backround is to grant access to only a subset of endpoints and subset of data those endpoints provide. 

I am unsure what would be the most elegant way to do this and would like to ask the community about ideas here.

**Comment by shashstormer** [comment]
I would suggest use middleware to filter allowed paths domains
```python
@app.middleware("*")
async def filter_requests(request: Request, next_process):
    if request.url.netloc not in ["example.net"] and (request.url.path.startswith("/api/...") and request.url.path != "/api/.../private"):
       #  This part of of code is run when the domain is not example.net and the url is like "/api/*" and the route is not  "/api/.../private"
        try:
            return await next_process(request)
        except RuntimeError:
            return Response("", status_code=404)
    if request.url.netloc not in ["example.com", "127.0.0.1:5010"]:
        # this part is returned when the above is false and the domain is not in the allowed domains
        return Response('{"details":{"error": 444}}', status_code=444)
   # this part is executed when the the domain is example.com or 127.0.0.1:5010
   return await next_process(request)
```

you may further customize your endpoint and permissions using this method

**Comment by aanchlia** [comment]
Have you thought about dividing in Sub Applications https://fastapi.tiangolo.com/advanced/sub-applications/ ?

**Comment by YuriiMotov** [comment, marked as answer]
Two more ways are:
* API-gateway
* [custom APIRoute class](https://fastapi.tiangolo.com/how-to/custom-request-and-route/)

Middleware and sub-applications mentioned above are also suitable options.

I would choose either sub-applications or API-gateway
